### ISSUE TITLE
Race condition when joining channels that require registration on Freenode

### ISSUE THREAD

**Issue by elazar** [phergie/phergie-irc-plugin-react-autojoin#5]
This plugin attempts to auto-join channels [after the MOTD has been received](https://github.com/phergie/phergie-irc-plugin-react-autojoin/blob/44f1aa40214dc0b0770188a0db092912d0d52932/src/Plugin.php#L79-80), whereas the [NickServ plugin](https://github.com/phergie/phergie-irc-plugin-react-nickserv) attempts to authenticate [when an appropriate notice is received](https://github.com/phergie/phergie-irc-plugin-react-nickserv/blob/9186aa7a25fa4d5fc7af7e970be3e3d41a917915/src/Plugin.php#L66). If the auto-join happens first, any attempts to join a channel on Freenode that requires registration will fail, as shown in the output below.

```
[2015-04-08/17:06:20]  irc.freenode.net <- 376 Phergie :End of /MOTD command.
[2015-04-08/17:06:20]  AutoJoin plugin: irc.freenode.net -> JOIN #phpc,#phpwomen,#phergie,#atlantaphp,##backroom,#phptek,#phpio,#nolainteractive,#exaviar,#pearanalytics_dev,##cincydev,#mta,#zendcon,#mswds,#phpc.de,#dallasphp,#mtrack,#spaz,#nashvillephp,#phpgroups,#btvphpug,#ottawaphp,#frapi,#phpsw,#web2project,#/r/dallas,#phpmentoring,#GuelphPHP,#conjunctured,#phpdiversity
[2015-04-08/17:06:20]  irc.freenode.net <- MODE Phergie :+i
[2015-04-08/17:06:20]  irc.freenode.net <- NOTICE Phergie :This nickname is registered. Please choose a different nickname, or identify via ^B/msg NickServ identify <password>^B.
[2015-04-08/17:06:20]  NickServ plugin: irc.freenode.net -> PRIVMSG NickServ IDENTIFY mgt37ge
[2015-04-08/17:06:26]  irc.freenode.net <- 470 Phergie #phpc ##please_register :Forwarding to another channel
[2015-04-08/17:06:26]  irc.freenode.net <- JOIN ##please_register
[2015-04-08/17:06:26]  irc.freenode.net <- 332 Phergie ##please_register :The channel you tried to join is set with mode +r. Only users who are registered ("/msg nickserv help register") and identified ("/msg nickserv help identify") to NickServ can join. Please follow the steps on http://freenode.net/faq.shtml#nicksetup .  Then "/part #please_register". Then /part the channel you first tried to join, and try again. Thanks!
[2015-04-08/17:06:26]  irc.freenode.net <- 333 Phergie ##please_register unknown 1212089057
```

`470` server responses where the last two arguments are `'##please_register'` and `'Forwarding to another channel'` need to invoke a second attempt to join the channel referenced immediately before those two arguments (e.g. in the above case, `#phpc`). This should be done in response to an event fired by the NickServ plugin when authentication is successful (see phergie/phergie-irc-plugin-react-nickserv#6).

**Comment by Renegade334** [comment]
One way to address this would be to have NickServ emit a "nickserv.identified" event on receiving the "you are now identified" notice, and for the AutoJoin plugin to take a config option for whether or not to wait for it.

**Comment by elazar** [comment]
@Renegade334 The NickServ event is why I mentioned phergie/phergie-irc-plugin-react-nickserv#6 above. The configuration option makes sense; it seems a better idea than trying to have AutoJoin figure out whether to wait for the MOTD or for NickServ.

**Comment by sitedyno** [comment]
Would it be simpler to have a config option to specify non-authenticated channels and a config option to specify authenticated channels? Then the user decides on the behavior and could even mix them if they wanted.

**Comment by Renegade334** [comment]
Certainly possible, although personally I think having it as a binary option would be fine. (7a0c037de82b24c4423289b8a308882795df3f94 is the implementation I had in mind.)

**Comment by elazar** [comment]
@sitedyno I'd rather avoid requiring the user to know ahead of time whether channels require authentication or not, since it's something we can feasibly determine at runtime, or to have to manually maintain two sets of channel lists.

**Comment by elazar** [comment]
phergie/phergie-irc-plugin-react-nickserv#6 has been addressed, so it should be possible to implement the solution discussed for this issue now.

**Comment by Renegade334** [comment]
I have a patch waiting to go, once a tag is dropped.

**Comment by elazar** [comment]
https://github.com/phergie/phergie-irc-plugin-react-nickserv/releases/tag/1.3.0 has been created.

**Comment by elazar** [comment]
Merged #6.
